### DISCUSSION TITLE
Scale recommendations for large servers

### DISCUSSION THREAD

**Discussion by candlerb** [seaweedfs/seaweedfs#1941] [Q&A] [answered]
I notice there is the "large disk" option which allows larger volumes (up to 8TB, rather than up to 32GB), so that the master server didn't have to keep track of large numbers of volumes.  I wondered if you had any recommendations for what would be a target range of volumes to aim for.

1. The specific application I have in mind is for around 10 servers, each holding 300TB, with an average file size of 15MB - i.e. about 20M files per server.

    The default 30GB volume size would give 10,000 volumes per server and up to 100,000 volumes in total.  Is that too large?  How much RAM is required per volume on the master server, and what are the other scaling limits?

    Otherwise, what would you recommend: 300GB volumes?  3TB volumes?  I can imagine there are other tradeoffs to consider, such as wasted space from part-filled volumes, and time spent repairing an individual volume.

2. I was thinking of making each server have 3 x 100TB RAID6 arrays.  Does that sound a reasonable strategy?  Would you normally recommend some sort of RAID underneath SeaweedFS volumes - or would it be better to serve each individual disk as a volume server, and use Seaweed's own replication instead of RAID?

My current thought is that 1TB volumes (representing 1% of each RAID6, and 300 per server) might be a reasonable setting.  If a 12-disk array can achieve 500MB/sec, then an operation that involves copying or rebuilding 1TB will take around half an hour and use about half of a 10G network connection.

**Comment by chrislusf** [comment, marked as answer]
Here are a few reasons about choosing volume size:
1. Less number of volumes is easier for master. 
2. It also depends on whether there are a lot of files updated or deleted, which will involve compaction and smaller volume is better.
3. One S3 bucket is mapped to a collection of volumes. More buckets need more volumes.
4. One Kubernetes persistent volume also maps to a collection of volumes.
5. Each variations of (collection, replication, TTL, diskType) needs one set of volumes.

Using RAID seems ok but not necessary. It only provides redundancy on one server. You would need cross server redundancy also, right? 

The hard drive is 25TB? or so? Keeping one volume server around 20 ~ 50 volumes should be good to start.

The max volume size limit can be changed. The volumes over the limit will just become readonly.

**Reply by candlerb** [reply]
> The hard drive is 25TB? or so?

12TB drives x 36 slots (Supermicro 4U: 24 at front and 12 at the back)

> Using RAID seems ok but not necessary. It only provides redundancy on one server. You would need cross server redundancy also, right?

For some use cases, a single copy with RAID6 underneath and no mirroring might be sufficient - where high availability is not required.  For that case I was thinking 3 x RAID6 (12 disks each), `weed volume -dir=/data1,/data2,/data3`

However, I'd like to consider the other case as well.

Suppose we abandoned the RAID6, and used cross-server mirroring for redundancy instead.  Would you suggest having 36 separate mount points, running with `weed volume -dir=/data{1..36}` - is that a practical deployment?

My concern is, suppose 1 of the 36 disks fails or locks up.  Is the master able to mark one disk as down, so that it marks only the volumes on that disk as down, without marking the whole volume server as down?

Also, I think that the volume size would have to be tweaked so it divides nicely into 12TB without too much wastage.  Perhaps 3900GB, which would also work if we later upgrade to 16TB drives.

> The max volume size limit can be changed. The volumes over the limit will just become readonly.

That is useful to know.  At least, I can see *reducing* the volume size limit is safe.

What happens if you *increase* the volume size limit?  Do existing open volumes still honour the size limit which was present at the time they were created, or could they grow further?  At least there is `-minFreeSpacePercent` as a safety limit.

**Reply by chrislusf** [reply]
You can create 36 volume servers, in case one disk fails.



> could they grow further

Yes

**Reply by candlerb** [reply]
> You can create 36 volume servers, in case one disk fails

That is, 36 separate instances of `weed volume`?  That's not too bad.
